crates/libs/core/src/event.rs

```diff
@@ -25,47 +29,52 @@ impl<T: Interface> Event<T> {
     /// Creates a new, empty `Event<T>`.
     pub fn new() -> Self {
         Self {
-            delegates: Array::new(),
+            delegates: UnsafeCell::new(Array::new()),
             swap: Mutex::default(),
             change: Mutex::default(),
         }
     }
 
     /// Registers a delegate with the event object.
-    pub fn add(&mut self, delegate: &T) -> Result<i64> {
+    pub fn add(&self, delegate: &T) -> Result<i64> {
         let mut _lock_free_drop = Array::new();
         Ok({
             let _change_lock = self.change.lock().unwrap();
-            let mut new_delegates = Array::with_capacity(self.delegates.len() + 1)?;
-            for delegate in self.delegates.as_slice() {
+            // Safety: there is no mutable alias to self.delegates at this point
```

Comment:
This comment says a condition is true, but I don't see how that is guaranteed.  I see calls to modify `delegates` in the `add` and `remove` paths, and I see a call that reads `delegates` in the `call` path.  But I don't see how this guarantees the safety invariant implied by this comment.

How are concurrent calls to `add` and `call` guaranteed to be safe?

Reply:
From my understanding exclusive (mutable) access is granted only by acquiring both locks. Though this is by convention rather then enforced. So acquiring one lock (which all functions do) prevents mutable access.

The pattern for `add`, `remove` and `clear` is to acquire a readonly `change` lock and then later upgrade it to an exclusive lock (by acquiring the `swap` lock too). The upgrade is conditional for `remove` and `clear`.

`call` simply acquires the `swap` lock, which blocks the other functions from upgrading their locks.

Reply:
Ok, thanks.  I'm going to study it a bit more and see if I agree.

Reply:
I see what you mean. My intention is when taking an immutable reference and reading from `add` (e.g. lines 43-48), we can still take an immutable reference at `call` as long as `add` haven't acquired the `self.swap` lock. I noticed that the `current_delegates` reference is still valid at line 55 (which is illegal) so I should've dropped it at line 49. This way `current_delegates` doesn't alias with line 55. Does this make sense?